xdsl/dialects/mesh.py

```diff
@@ -1,18 +1,99 @@
+from collections.abc import Sequence
+from typing import ClassVar, TypeAlias
+
 from xdsl.dialects.builtin import (
+    I16,
     I64,
+    ArrayAttr,
+    BytesAttr,
     DenseArrayBase,
     SymbolNameConstraint,
+    i16,
 )
 from xdsl.dialects.utils.dimension_list import DimensionList
-from xdsl.ir import Dialect, VerifyException
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    OpaqueSyntaxAttribute,
+    ParametrizedAttribute,
+    VerifyException,
+)
 from xdsl.irdl import (
     IRDLOperation,
+    irdl_attr_definition,
     irdl_op_definition,
     prop_def,
     traits_def,
 )
+from xdsl.parser import AttrParser
+from xdsl.printer import Printer
 from xdsl.traits import SymbolOpInterface
 
+MeshAxesAttr: TypeAlias = DenseArrayBase[I16]
+
+
+@irdl_attr_definition
+class MeshAxesArrayAttr(ParametrizedAttribute, OpaqueSyntaxAttribute):
+    """
+    MeshAxesArrayAttr attribute for representing mutiple mesh axes.
+
+    Reflects [the MLIR attribute](https://github.com/llvm/llvm-project/blob/6146a88f60492b520a36f8f8f3231e15f3cc6082/mlir/include/mlir/Dialect/Mesh/IR/MeshBase.td#L83)
+    """
+
+    name = "mesh.axisarray"
+
+    axes: ArrayAttr[MeshAxesAttr]
+
+    ELT_TYPE: ClassVar = i16
+
+    @classmethod
+    def _parse_mesh_axes_attr(cls, parser: AttrParser) -> DenseArrayBase[I16]:
+        """
+        Parses a single MeshAxesAttr, e.g. [1, 4, 7, 8]
+        """
+        elements = parser.parse_comma_separated_list(
+            parser.Delimiter.SQUARE,
+            parser.parse_integer,
+        )
+
+        return DenseArrayBase[I16](cls.ELT_TYPE, BytesAttr(cls.ELT_TYPE.pack(elements)))
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        """
+        Parses a MeshAxesArrayAttr, which has the syntax of a list
+        of lists, e.g.:
+
+        [[1, 2, 3], [], [4, 5]]
+        """
+        axes = parser.parse_comma_separated_list(
+            parser.Delimiter.SQUARE,
+            lambda: cls._parse_mesh_axes_attr(parser),
+        )
+
+        return (ArrayAttr(axes),)
+
+    @classmethod
+    def _print_sublist(cls, sublist: MeshAxesAttr, printer: Printer) -> None:
```

Comment:
I have a tiny preference for printer/parser as first arg in these helpers but it's not a big deal